profiling/paper_timing_tables/timing_comparison.py

```diff
@@ -0,0 +1,338 @@
+from astropy.io import ascii
+
+MAXIT = 1  # number of iterations to time and average
+# number of TOAs run for each test
+ntoas_simple = [100, 1000, 10000, 100000]
+ntoas_complex = [5012, 10024, 25060]
+par_simple = "NGC6440E.par"
+timfiles = [
+    "NGC6440E_fake100.tim",
+    "NGC6440E_fake1000.tim",
+    "NGC6440E_fake10000.tim",
+    "NGC6440E_fake100000.tim",
+]  # timfiles for simple model and individual functions
+timfiles2 = [
+    "J1910+1256_NANOGrav_12yv4.tim",
+    "J1910+1256_NANOGrav_12yv4_10k.tim",
+    "J1910+1256_NANOGrav_12yv4_25k.tim",
+]  # timfiles for complex model
+
+
+def pintrun(parfile, timfile, ptime_arr, pickle, fitter):
+    """ Runs and times pintempo 5 times and averages times, appending to a list. """
+    gls = ""
+    if fitter == "gls":
+        gls = " --gls"
+    usepickle = ""
+    if pickle:
+        usepickle = " --usepickle"
+    total = 0
+    for i in range(MAXIT):
+        start = timeit.default_timer()
+        subprocess.check_call(
+            "pintempo" + usepickle + gls + " " + parfile + " " + timfile,
+            shell=True,
+            stdout=subprocess.DEVNULL,
+            stderr=subprocess.DEVNULL,
+        )
+        end = timeit.default_timer()
+        total = total + (end - start)
+    ptime_arr.append(total / MAXIT)  # averages time
+
+
+def temporun(parfile, timfile, ttime_arr, fitter):
+    """ Runs and times TEMPO 5 times and averages times, appending to a list. """
+    fit = ""
+    if fitter == "gls":
+        fit = " -G"
+    total = 0
+    for i in range(MAXIT):
+        start = timeit.default_timer()
+        subprocess.check_call(
+            "tempo" + fit + " -f " + parfile + " " + timfile + "",
+            shell=True,
+            stdout=subprocess.DEVNULL,
+            stderr=subprocess.DEVNULL,
+        )
+        end = timeit.default_timer()
+        total = total + (end - start)
+    ttime_arr.append(total / MAXIT)  # average time
+
+
+def tempo2run(parfile, timfile, t2time_arr):
+    """ Runs and times Tempo2 5 times and averages times, appending to a list. """
+    total = 0
+    for i in range(MAXIT):
+        start = timeit.default_timer()
+        subprocess.check_call(
+            "tempo2 -nobs 100003 -f " + parfile + " " + timfile + "",
+            shell=True,
+            stdout=subprocess.DEVNULL,
+            stderr=subprocess.DEVNULL,
+        )
+        end = timeit.default_timer()
+        total = total + (end - start)
+    t2time_arr.append(total / MAXIT)  # average time
+
+
+if __name__ == "__main__":
+    with TemporaryDirectory() as tempdir:
+        # Generate simple, fake TOAs for the timing runs
+        print("Making fake TOAs...")
+        for num in ntoas_simple:
+            call = (
+                "zima --startMJD 53478 --duration 700 --freq 1400 2000 --ntoa "
+                + str(num)
+                + " "
+                + par_simple
+                + " "
+                + tempdir
+                + "/NGC6440E_fake"
```

Comment:
It's a good idea to use `os.path.join()` to join paths rather than concatenating with a '/'; there are weird corner cases (and Windows) where `join()` works but that doesn't. It may be worth a `from os.path import join` if you use it a lot.